bundle/config/validate/single_node_cluster.go

```diff
@@ -0,0 +1,174 @@
+package validate
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/databricks/cli/bundle"
+	"github.com/databricks/cli/libs/diag"
+	"github.com/databricks/cli/libs/dyn"
+	"github.com/databricks/databricks-sdk-go/service/compute"
+	"github.com/databricks/databricks-sdk-go/service/pipelines"
+)
+
+// Validates that any single node clusters defined in the bundle are correctly configured.
+func SingleNodeCluster() bundle.ReadOnlyMutator {
+	return &singleNodeCluster{}
+}
+
+type singleNodeCluster struct{}
+
+func (m *singleNodeCluster) Name() string {
+	return "validate:SingleNodeCluster"
+}
+
+const singleNodeWarningDetail = `num_workers should be 0 only for single-node clusters. To create a
+valid single node cluster please ensure that the following properties
+are correctly set in the cluster specification:
+
+  spark_conf:
+    spark.databricks.cluster.profile: singleNode
+    spark.master: local[*]
+
+  custom_tags:
+    ResourceClass: SingleNode
+  `
+
+const singleNodeWarningSummary = `Single node cluster is not correctly configured`
+
+func validateSingleNodeCluster(spec *compute.ClusterSpec, l []dyn.Location, p dyn.Path) *diag.Diagnostic {
+	if spec == nil {
+		return nil
+	}
+
+	if spec.NumWorkers > 0 || spec.Autoscale != nil {
+		return nil
+	}
+
+	if spec.PolicyId != "" {
+		return nil
+	}
+
+	invalidSingleNodeWarning := &diag.Diagnostic{
+		Severity:  diag.Warning,
+		Summary:   singleNodeWarningSummary,
+		Detail:    singleNodeWarningDetail,
+		Locations: l,
+		Paths:     []dyn.Path{p},
+	}
+	profile, ok := spec.SparkConf["spark.databricks.cluster.profile"]
+	if !ok {
+		return invalidSingleNodeWarning
+	}
+	master, ok := spec.SparkConf["spark.master"]
+	if !ok {
+		return invalidSingleNodeWarning
+	}
+	resourceClass, ok := spec.CustomTags["ResourceClass"]
+	if !ok {
+		return invalidSingleNodeWarning
+	}
+
+	if profile == "singleNode" && strings.HasPrefix(master, "local") && resourceClass == "SingleNode" {
+		return nil
+	}
+
+	return invalidSingleNodeWarning
+}
+
+func validateSingleNodePipelineCluster(spec pipelines.PipelineCluster, l []dyn.Location, p dyn.Path) *diag.Diagnostic {
```

Comment:
You can avoid 2 similar functions by introducing new struct `singleNodeClusterSpec` containing fields you need and then validate it, so it becomes

```
func validateSingleNodeCluster(spec *singleNodeClusterSpec) bool

func fromJobSpec(...) *singleNodeClusterSpec
func fromPipelineSpec(...) *singleNodeClusterSpec
```

Reply:
Moved the warning to outside this function! New struct is no longer needed.